pkg/convert/jfr.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"context"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/google/pprof/profile"
+	"github.com/pyroscope-io/jfr-parser/parser"
+)
+
+type builder struct {
+	profile         *profile.Profile
+	locationTable   map[string]*profile.Location
+	functionTable   map[string]*profile.Function
+	sampleTable     map[string]*profile.Sample
+	sampleTypeTable map[string]int
+}
+
+func newBuilder() *builder {
+	return &builder{
+		profile:       &profile.Profile{SampleType: []*profile.ValueType{{Type: "cpu", Unit: "samples"}}},
+		locationTable: map[string]*profile.Location{},
+		functionTable: map[string]*profile.Function{},
+		sampleTable:   map[string]*profile.Sample{},
+	}
+}
+
+func JfrToPprof(ctx context.Context, r io.Reader) (*profile.Profile, error) {
+	chunks, err := parser.Parse(r)
+	if err != nil {
+		return nil, err
+	}
+
+	b := newBuilder()
+	for _, c := range chunks {
+		if err := b.addJFRChunk(ctx, c); err != nil {
+			return nil, err
+		}
+	}
+
+	return b.profile, nil
+}
+
+func (b *builder) addJFRChunk(ctx context.Context, c parser.Chunk) error {
+	var event string
+	for _, e := range c.Events {
+		if as, ok := e.(*parser.ActiveSetting); ok {
+			// What is this for!?
+			if as.Name == "event" {
+				event = as.Value
+			}
+		}
+	}
+	if event != "cpu" {
+		return nil
+	}
+
+	for _, event := range extractExecutionSampleEvents(c.Events) {
+		if event.State.Name == "STATE_RUNNABLE" {
+			increaseSample(b.getOrCreateSample(event.StackTrace))
```

Comment:
This is fine for now. And probably you've already thought about this but I think we'll need to keep a cache of stack traces that have already been seen so that we don't end up with duplicate samples. I suppose better to handle that when we call `addJFRChunk`?

Reply:
that cache already exists within the builder, if we've already seen a stack then the previously created sample is returned